### ISSUE TITLE
Allow configuration of encodings (fix Unicode issues)

### ISSUE THREAD

**Issue by mkleehammer** [mkleehammer/pyodbc#112]
There are a lot of Unicode issues due to the fact that drivers deal with SQLCHAR and SQLWCHAR in different ways.  If I understand the ODBC specification correctly, there is only one correct way which is:
- Use SQLCHAR for ANSII data
- Use SQLWCHAR for Unicode data, encoded using UCS-2 or UTF-16

Many drivers return Unicode data via SQLCHAR encoded using UTF-8.

I am proposing adding the ability to specify the encoding for SQLCHAR and SQLWCHAR separately.  Once done, we can gather up the recommended settings for different drivers.

I am planning on adding this to the Connection, allowing different instances to be configured differently.  To simplify usage, I'll also add a pair of global settings to the pyodbc module so you don't have to configure them individually.

Here is my initial proposal:

```
pyodbc.connect(..., char_encoding='utf-8', wchar_encoding='utf-16')
```

The names are a bit long but I think spelling out "encoding" matches other APIs.  Does anyone have other examples?

The values can be any accepted Python encoding, though I might consider using the built-in UTF-8 and UTF-16 APIs.

Perhaps these should also be the default values.

This is open for discussion and recommendations.

**Comment by rclough** [comment]
I'm not familiar enough with the gritty details for a valuable critique, but from an end users perspective, this solution would work fine for me (and is the kind of thing I'd have looked for when I first ran into my issue)

**Comment by Lexcon** [comment]
SQL Server has varchar and nvarchar versus text and ntext. nvarchar and ntext are obviously unicode but the others contain user defined content for which driver interpretation can be facilitating but not mandatory. Therefore, Varchar should be returned as a bytearray by default. Consider these records, where the encoding is part of the per-row information:

encoding        content 
windows-1250  Ďjordje             
windows-1252  Daniël

When I do 'select * from persons' I would like to be able to decode each row based on my own logic on Python level.

In practise I have seen varchar records that have all kinds of stuff in them, up to zipped strings (so binary).

**Comment by mkleehammer** [comment]
I've created a Unicode label for these issues.  I believe this issue would solve those.  (Drivers.  Sigh.)

**Comment by mkleehammer** [comment]
Now on the v4 branch.  I haven't written up docs yet, but the API changes are:

* add `encoding` keyword to connect to control encoding of connection string
* add setencoding to control encoding of SQL statements and parameters
* add setdecoding to control reading of SQL_CHAR and SQL_WCHAR columns

For Python 3, the defaults are:

    cnxn.setencoding('utf-16le')
    cnxn.setdecoding(pyodbc.SQL_CHAR, encoding='utf-8')
    cnxn.setdecoding(pyodbc.SQL_WCHAR, encoding='utf-16le')

Python 2 is similar but there are two encodings since there are two "text" types:

    cnxn.setencoding(str, 'utf-8')
    cnxn.setencoding(unicode, 'utf-16le')
    cnxn.setdecoding(pyodbc.SQL_CHAR, encoding='utf-8')
    cnxn.setdecoding(pyodbc.SQL_WCHAR, encoding='utf-16le')

**Comment by mkleehammer** [comment]
I've rebased the v4 branch so I could make a smaller pull request #160.

**Comment by Lexcon** [comment]
Thanks, sounds great,

Does it have by any chance have something like Connection.setdecoding('raw')

for just passing through bytes stored in varchars?

Robert

**Comment by mkleehammer** [comment]
I'm going to reply on the pull request so others can contribute.

**Comment by mkleehammer** [comment]
I believe Unicode issues are fixed in the newly released 4.0.0 version.
